Implement file database for indexing and querying

hendrikvanantwerpen opened the pull request:
| ◀️ #249 | #248 ▶️ |
|-|-|

_#248 should be merged as soon as possible after this one to prevent CI for other PRs based of `main` to be broken._

This PR is part of the work to implement an LSP server (#242). This implements file storage for stack graphs and partial paths, allowing separate `index` and `query` invocations.

## Changes

- A `storage` module in stack-graphs has the reader and writer types for the SQLite database we use. The reader supports piecemeal loading of paths into the runtime structures. It is guarded by the `storage` Cargo feature.
- A new `index` command replaces the `analyze` command. It takes a database argument where the results are stored.
- A new `query` command can be used to query the database. It currently has a single subcommand `definition` to query for definitions.
- A new `clean` command can be used to clean data from the database. It either expects source paths, in which case all descendants of these paths are cleaned, or the `--all` flag, in which case the whole database is cleaned.
- The database can store a tag for each file, which is used to implement incremental indexing. If the file content did not change, the file is not reindexed (unless `--force` is passed).

## Demo

<details>
<summary>First index</summary>

    me@localhost:~/src/typeorm/typeorm(master)$ tree-sitter-stack-graphs-typescript index -v --hide-error-details tsconfig.json package.json src/data-source/ src/entity-manager/
    /Users/me/src/typeorm/typeorm/tsconfig.json: not supported
    /Users/me/src/typeorm/typeorm/package.json: not supported
    /Users/me/src/typeorm/typeorm/src/data-source/BaseDataSourceOptions.ts: success
    /Users/me/src/typeorm/typeorm/src/data-source/DataSource.ts: success
    /Users/me/src/typeorm/typeorm/src/data-source/DataSourceOptions.ts: success
    /Users/me/src/typeorm/typeorm/src/data-source/index.ts: success
    /Users/me/src/typeorm/typeorm/src/entity-manager/EntityManager.ts: success
    /Users/me/src/typeorm/typeorm/src/entity-manager/EntityManagerFactory.ts: success
    /Users/me/src/typeorm/typeorm/src/entity-manager/MongoEntityManager.ts: success
    /Users/me/src/typeorm/typeorm/src/entity-manager/SqljsEntityManager.ts: success

</details>

<details>
<summary>Cached index</summary>

    me@localhost:~/src/typeorm/typeorm(master)$ tree-sitter-stack-graphs-typescript index -v --hide-error-details tsconfig.json package.json src/data-source/ src/entity-manager/
    /Users/me/src/typeorm/typeorm/tsconfig.json: not supported
    /Users/me/src/typeorm/typeorm/package.json: not supported
    /Users/me/src/typeorm/typeorm/src/data-source/BaseDataSourceOptions.ts: cached
    /Users/me/src/typeorm/typeorm/src/data-source/DataSource.ts: cached
    /Users/me/src/typeorm/typeorm/src/data-source/DataSourceOptions.ts: cached
    /Users/me/src/typeorm/typeorm/src/data-source/index.ts: cached
    /Users/me/src/typeorm/typeorm/src/entity-manager/EntityManager.ts: cached
    /Users/me/src/typeorm/typeorm/src/entity-manager/EntityManagerFactory.ts: cached
    /Users/me/src/typeorm/typeorm/src/entity-manager/MongoEntityManager.ts: cached
    /Users/me/src/typeorm/typeorm/src/entity-manager/SqljsEntityManager.ts: cached

</details>

<details>
<summary>Forced index</summary>

    me@localhost:~/src/typeorm/typeorm(master)$ tree-sitter-stack-graphs-typescript index -v --hide-error-details --force tsconfig.json package.json src/data-source/ src/entity-manager/
    /Users/me/src/typeorm/typeorm/tsconfig.json: not supported
    /Users/me/src/typeorm/typeorm/package.json: not supported
    /Users/me/src/typeorm/typeorm/src/data-source/BaseDataSourceOptions.ts: success
    /Users/me/src/typeorm/typeorm/src/data-source/DataSource.ts: success
    /Users/me/src/typeorm/typeorm/src/data-source/DataSourceOptions.ts: success
    /Users/me/src/typeorm/typeorm/src/data-source/index.ts: success
    /Users/me/src/typeorm/typeorm/src/entity-manager/EntityManager.ts: success
    /Users/me/src/typeorm/typeorm/src/entity-manager/EntityManagerFactory.ts: success
    /Users/me/src/typeorm/typeorm/src/entity-manager/MongoEntityManager.ts: success
    /Users/me/src/typeorm/typeorm/src/entity-manager/SqljsEntityManager.ts: success

</details>

<details>
<summary>Clean and re-index</summary>

    me@localhost:~/src/typeorm/typeorm(master)$ tree-sitter-stack-graphs-typescript clean .

    me@localhost:~/src/typeorm/typeorm(master)$ tree-sitter-stack-graphs-typescript index -v --hide-error-details tsconfig.json package.json src/data-source/ src/entity-manager/
    /Users/me/src/typeorm/typeorm/tsconfig.json: not supported
    /Users/me/src/typeorm/typeorm/package.json: not supported
    /Users/me/src/typeorm/typeorm/src/data-source/BaseDataSourceOptions.ts: success
    /Users/me/src/typeorm/typeorm/src/data-source/DataSource.ts: success
    /Users/me/src/typeorm/typeorm/src/data-source/DataSourceOptions.ts: success
    /Users/me/src/typeorm/typeorm/src/data-source/index.ts: success
    /Users/me/src/typeorm/typeorm/src/entity-manager/EntityManager.ts: success
    /Users/me/src/typeorm/typeorm/src/entity-manager/EntityManagerFactory.ts: success
    /Users/me/src/typeorm/typeorm/src/entity-manager/MongoEntityManager.ts: success
    /Users/me/src/typeorm/typeorm/src/entity-manager/SqljsEntityManager.ts: success

</details>

<details>
<summary>Query definition</summary>

    me@localhost:~/src/typeorm/typeorm(master)$ tree-sitter-stack-graphs-typescript query definition src/data-source/DataSource.ts 84 23
    /Users/me/src/typeorm/typeorm/src/data-source/DataSource.ts:84:23: found definitions:
       0: /Users/me/src/typeorm/typeorm/src/data-source/DataSource.ts:6:10

</details>

hendrikvanantwerpen commented:
I am very surprised by the change in memory usage. I don't really know how to explain that 🤔

rnkaufman commented on tree-sitter-stack-graphs/src/cli/index.rs:
```diff
@@ -86,57 +92,69 @@ impl AnalyzeArgs {
-                    self.analyze_file(source_root, source_path, loader, &mut seen_mark)?;
+                    let source_path = source_entry.path().canonicalize()?;
+                    self.analyze_file(
+                        source_root,
+                        &source_path,
```
Is it intended that only one of these is a reference? Same for lines 128-129

hendrikvanantwerpen replied:
Good question! Both were passed by reference, but `source_root` was already a reference, while `source_path` was a value. However, Rick made a nice suggestion above that allows using `source_path` now.

rewinfrey commented on stack-graphs/src/storage.rs:
```diff
@@ -0,0 +1,472 @@
+    }
+
+    /// Open a file database.  If the file does not exist, it is automatically created.
+    /// An error is thrown if the database version is not supported.
```
Small nit: Could this be updated to say `An error is returned ...` instead of `throws`?

rewinfrey commented on stack-graphs/src/storage.rs:
```diff
@@ -0,0 +1,472 @@
+                "added path {} must start in given file {} or at root",
+                path.display(graph, partials),
+                graph[file].name()
+            );
```
General question: when a function can `panic`, should that be indicated in the comment for the function? I'm not sure if there is a Rust convention for that, and this isn't intended to be a suggestion here, it's only a drive-by question.

hendrikvanantwerpen replied:
Good question. I think you're right that it is a good idea to add that in the comment!

rewinfrey reviewed (commented):
I think I have more questions than comments or suggestions 😄, but left a couple minor things for you to consider. I think the SQLite implementation and additional database commands look good to me.

hendrikvanantwerpen commented on stack-graphs/src/assert.rs:
```diff
@@ -72,6 +72,22 @@ impl AssertionSource {
+            }
+        }
+        Displayer(self, graph)
+    }
```
You might be right, that we could capture the graph in the `AssertionSource` itself. Without having looked at all the current uses, I think this might make it a lot less flexible to use, though. As I recall, these are constructed from tests _before_ the graph even exists (and their construction does not need a graph either). So capturing the graph earlier would restrict users quite a bit.

In one of the later following PRs I've dropped the use of `AssertionSource` in the CLI again, because it requires full LSP positions which require reading the source file to be calculated.

hendrikvanantwerpen replied:
I looked in the source, and it is not exactly as I said. The graph must already exist, because we store a `Handle<File>` in `AssertionSource`. But I think my point is still valid, because capturing the graph would e.g. not allow any `&mut` references as long as the `AssertionSource` exists. In other words, it would still force a user to have the graph fully constructed before creating `AssertionSource`s.

rewinfrey replied:
That's a great point that having a reference to the `graph` in `AssertionSource` would complicate `&mut` references elsewhere. I still think the current solution is 👍 and a nice way to handle this.

hendrikvanantwerpen commented on stack-graphs/src/stitching.rs:
```diff
@@ -485,54 +485,32 @@ impl ForwardPartialPathStitcher {
         graph: &StackGraph,
         partials: &mut PartialPaths,
-        db: &mut Database,
+        _db: &mut Database,
```
Initially it was there for consistency with the `from_partial_paths` method, which _did_ use it. However, that was mostly because the constructors were already doing a bit of work. I've changed that at some point so that the constructor only sets up the stitcher, but doesn't do work until `process_next_phase` is called. Given that neither of the constructors use `db` anymore, I think we can remove it.

hendrikvanantwerpen commented on tree-sitter-stack-graphs/src/cli/database.rs:
```diff
@@ -0,0 +1,42 @@
+}
+
+impl DatabaseArgs {
+    pub fn default_for_crate(crate_name: &str) -> anyhow::Result<PathBuf> {
```
This method is perhaps a bit confusing here, since it does not actually create or act upon `DatabaseArgs` values. It's an associated method to easily get a default database location for a given crate name. I'll move this to be a module level function instead, so that it's clearer that it does not use `DatabaseArgs` in any way.

hendrikvanantwerpen commented on tree-sitter-stack-graphs/src/cli/index.rs:
```diff
@@ -86,57 +92,69 @@ impl AnalyzeArgs {
         let mut seen_mark = false;
+        let mut db = SQLiteWriter::open(&db_path)?;
         for source_path in &self.source_paths {
+            let source_path = source_path.canonicalize()?;
```
How about that, that works :). That's a bit cleaner!

I thought that the value created by `canonicalize()` would not live long enough if you immediately take the reference but don't capture the value itself as well. Here it is accepted, although I know sometimes it doesn't work. Now I'm not sure when it is or is not allowed 🤔.

hendrikvanantwerpen commented:
@rewinfrey & @rnkaufman, thank you for taking a look at this. I think I addressed all your comments.

I've included one more change, which changes the command-line syntax for queries. Before, a query was `query definition PATH LINE COLUMN`. Now it is `query definition PATH:LINE:COLUMN+`.

I've done this for two reasons:

1. Multiple queries can be combined in one invocation. This is more flexible for the user, and might make it easier to do some automated profiling later.
2. This format is closer to the format we and many other tools use when printing source locations. This will make copy/pasting easier.

rewinfrey reviewed (approved):
👍 this looks great to me, thank you for the additional context and the answers 😄
